storage/backup/backup-cli/src/coordinators/backup.rs

```diff
@@ -328,6 +328,76 @@ impl BackupCoordinator {
     }
 }
 
+pub struct BackupCompactor {
+    storage: Arc<dyn BackupStorage>,
+    metadata_cache_opt: MetadataCacheOpt,
+    epoch_ending_file_compact_cnt: usize,
+    state_snapshot_file_compact_cnt: usize,
+    transaction_file_compact_cnt: usize,
+    concurrent_downloads: usize,
+}
+
+impl BackupCompactor {
+    pub fn new(
+        epoch_ending_file_compact_cnt: usize,
+        state_snapshot_file_compact_cnt: usize,
+        transaction_file_compact_cnt: usize,
+        metadata_cache_opt: MetadataCacheOpt,
+        storage: Arc<dyn BackupStorage>,
+        concurrent_downloads: usize,
+    ) -> Self {
+        BackupCompactor {
+            storage,
+            metadata_cache_opt,
+            epoch_ending_file_compact_cnt,
+            state_snapshot_file_compact_cnt,
+            transaction_file_compact_cnt,
+            concurrent_downloads,
+        }
+    }
+
+    pub async fn run(&self) -> Result<()> {
+        info!("Backup compaction started");
+        // Connect to both the local node and the backup storage.
+        let mut metaview = metadata::cache::sync_and_load(
+            &self.metadata_cache_opt,
+            Arc::clone(&self.storage),
+            self.concurrent_downloads,
+        )
+        .await?;
+
+        // Get a snapshot of remote metadata files before compaction
+        let files = self.storage.list_metadata_files().await?;
+
+        // Get all compacted chunks to be written back to storage
+        for range in metaview.compact_epoch_ending_backups(self.epoch_ending_file_compact_cnt)? {
+            let epoch_range = Metadata::new_epoch_ending_backup_range(range.to_vec())?;
+            self.storage
+                .save_metadata_line(&epoch_range.name(), &epoch_range.to_text_line()?)
+                .await?;
+        }
```

Comment:
A client can see a same txn in two different txn backups, for example, while you are doing this. I assume you'll deal with that in a separate PR?

Reply:
good catch. add the sort_unstable and dedup in the metaview part